### PR TITLE
implementation of linked list and tests

### PR DESCRIPTION
**Pull request by florian-kloeck**

### CODE REVIEW

File: course4/homework-linked-list/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,10 @@
+cmake_minimum_required(VERSION 3.0 FATAL_ERROR)
+
+set(CMAKE_CXX_STANDARD 14)
```
Comment (maciekgajewski) [inline course4/homework-linked-list/CMakeLists.txt:3]: This is nicer than the explicit setting of params that I usually do. Unfortunately, it doesn't enable warnings.

File: course4/homework-linked-list/linked_list.hh [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+namespace course {
+
+template<typename T>
+struct pointer_value_pair
+{
+	pointer_value_pair(std::unique_ptr<pointer_value_pair<T>>&& pointer, T& value) : value(value), pointer(std::move(pointer)) {}
+	
+	std::unique_ptr<pointer_value_pair<T>> pointer;
+    T& value;
```
Comment (maciekgajewski) [inline course4/homework-linked-list/linked_list.hh]: Something's wrong with indentation here

File: course4/homework-linked-list/test.cc [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+	assert(std::find_if(linked_list.begin(), linked_list.end(), [&](int n){return n>=i1;}) == linked_list.begin());
+	assert(*std::find_if(linked_list.begin(), linked_list.end(), [&](int n){return n>=i2;}) == ((i2>i1)?i2:i1));
+	assert(std::find_if(linked_list.begin(), linked_list.end(), [&](int n){return n>std::max(i1, i2);}) == linked_list.end());
+	assert(empty.begin()==empty.end());
+	assert(std::find_if(empty.begin(), empty.end(), [&](int n){return n>std::max(i1, i2);}) == empty.end());
+
+	//changing the value in the front of the list
+	linked_list.front() = i2;
+	assert(linked_list.front() == i2);
+}
```
Comment (maciekgajewski) [inline course4/homework-linked-list/test.cc:87]: Please, try the following test: fill the list with 100 numbers, in a loop, calculate the sum using std::accumulate

**Review by maciekgajewski — COMMENTED** [review]
Your list, while mosty good, is not really a container.
Please, expand the tests as suggested.

**Review by maciekgajewski — COMMENTED** [review]
Good!

File: course4/homework-linked-list/count_until.hh [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+#include <iterator>
+
+namespace course {
+
+template<class InputIt, class T>
+typename std::iterator_traits<InputIt>::difference_type
+    count_until(InputIt first, InputIt last, const T& value, const T& stop)
+{
+    typename std::iterator_traits<InputIt>::difference_type ret = 0;
+    for (; (first != last) && (*first)!=stop; ++first) {
```
Comment (maciekgajewski) [inline course4/homework-linked-list/count_until.hh:12]: Can you make it _without_ using the for-loop? (or any other loop)

**Review by maciekgajewski — COMMENTED** [review]
Try implementing the algo w/o explicit loop

File: course4/homework-linked-list/count_until.hh (hunk 4 above)
Comment (florian-kloeck) [reply course4/homework-linked-list/count_until.hh:12]: I can think of two solutions:
- std::count(first, std::find(first, last, stop), value);
- or using std::count_if with a predicate that returns false after observing the stop value

but both of these have worse complexity than this explicit loop. Is there a solution with the same complexity as the loop above?

Comment (maciekgajewski) [reply course4/homework-linked-list/count_until.hh:12]: The first solution looks very nice,m although, indeed, int will iterate over the collection twice.

Why do you think that the second one would be worse?

Comment (florian-kloeck) [reply course4/homework-linked-list/count_until.hh:12]: Because you don't stop iterating after finding the stop value

**Pull request closed without merging**